controllers/apim/ratelimitpolicy_controller.go

```diff
@@ -108,61 +105,128 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
 		return ctrl.Result{}, err
 	}
 
-	for _, networkingRef := range rlp.Spec.NetworkingRef {
-		switch networkingRef.Type {
-		case apimv1alpha1.NetworkingRefTypeHR:
-			logger.Info("HTTPRoute is not implemented yet") // TODO(rahulanand16nov)
-			return ctrl.Result{}, nil
-		case apimv1alpha1.NetworkingRefTypeVS:
-			vsNamespacedName := client.ObjectKey{
-				Namespace: rlp.Namespace, // VirtualService lookup is limited to RLP's namespace
-				Name:      networkingRef.Name,
-			}
+	// Operation specific annotations must be removed if they were present
+	updateRequired := false
+	// check for delete operation for virtualservice
+	if vsName, present := rlp.Annotations[KuadrantDeleteVSAnnotation]; present {
+		vsNamespacedName := client.ObjectKey{
+			Namespace: rlp.Namespace, // VirtualService lookup is limited to RLP's namespace
+			Name:      vsName,
+		}
+		vsKey := vsNamespacedName.String()
 
-			var vs istio.VirtualService
-			if err := r.Client().Get(ctx, vsNamespacedName, &vs); err != nil {
-				if apierrors.IsNotFound(err) {
-					logger.Info("no VirtualService found", "lookup name", vsNamespacedName.String())
-					return ctrl.Result{}, nil
-				}
-				logger.Error(err, "failed to get VirutalService")
-				return ctrl.Result{}, err
+		var vs istio.VirtualService
+		if err := r.Client().Get(ctx, vsNamespacedName, &vs); err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.Info("no VirtualService found", "lookup name", vsNamespacedName.String())
+				return ctrl.Result{}, nil
 			}
+			logger.Error(err, "failed to get VirutalService")
+			return ctrl.Result{}, err
+		}
+
+		if err := r.detachFromNetwork(ctx, vs.Spec.Gateways, vsKey, &rlp); err != nil {
```

Comment:
I think with the annotation based attachment, we would need a VS controller that watched VS. When the VS is deleted, it would call into a ratelimiting package to detach and an auth package. If the rl annotation is removed, then the VS controller reconciler would be called and it would again call to the ratelimiting package to detach. @eguzki @rahulanand16nov WDYT?

Reply:
Yup, it was part of the consideration and we can totally do that. The only opposite point was that the RLP controller does nothing but create limitador's `RateLimit` resources. The reasoning behind it was to separate the concerns related to ratelimit to the RLP controller only. If you and @eguzki think it's the better way to go then I'll move the code.

Reply:
Plus, I was thinking RLP can store in its status block about various networking object that particular RLP is "attached" to.

Reply:
> I think with the annotation based attachment, we would need a VS controller that watched VS. When the VS is deleted, it would call into a ratelimiting package to detach and an auth package. If the rl annotation is removed, then the VS controller reconciler would be called and it would again call to the ratelimiting package to detach. @eguzki

But, we do have a VS controller listening to events of the VS resources. https://github.com/Kuadrant/kuadrant-controller/pull/86/files#diff-9657787be5bdaae1595dcc9da8bd48808e19334779f7deeab86dc912d5d65478L181 Is that what you were asking for? 

Some context about this:

The reconciled resources, EnvoyFilters and RateLimits, needs data from both the RLP and the VirtualService (Host and gateway(s) namespace and names). As I see, both approaches have sense, i.e., implementing the reconciliation job from the VS controller or the RLP controller. 

However, I think we need to distinguish between reconciling EnvoyFilters and RateLimits and catching events of VS deletion, VS annotation update/delete, RLP update and so on. Regarding the VS events, it seems reasonable to be done in the VS controllers, whereas regarding thef RLP events, it should be done by the RLP controller. The reconciliation of the EnvoyFilters and RateLimits could be implemented in both controllers.  And why not reconciling EnvoyFilters and RateLimits when catching events? Let me try to explain as I see:

In the operator-framework controller pattern, the reconciliation job is done in the reconciler object. Each controller has one reconciler main type (assigned with the `builder.For(TYPE)` method). Besides the main type, each controller can register [event handlers](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.10.0/pkg/handler/eventhandler.go#L42)  for other types as well. These event handlers are meant for (and I quote):
```
// EventHandler enqueues reconcile.Requests in response to events (e.g. Pod Create).  EventHandlers map an Event 
// for one object to trigger Reconciles for either the same object or different objects - e.g. if there is an    
// Event for object with type Foo (using source.KindSource) then reconcile one or more object(s) with type Bar.  
```

Thus, I feel that is not right to do the reconciliation job (creating, updating, deleting reconciled resources based on desired state) from within the event handlers. I think they were not designed for that. One clear sign of that is the error handling of the event handler. There is not. So, basically you cannot tell to retry.